core/shared/src/main/scala/scalaz/zio/Promise.scala

```diff
@@ -53,6 +53,22 @@ class Promise[E, A] private (private val state: AtomicReference[State[E, A]]) ex
       result
     })
 
+  /**
+   * Retrieves the value of the promise, inside an Option,
+   * if it is immediately available
+   * Otherwise completes  immediately with None
+   */
+  final def peek: IO[E, Option[A]] =
+    IO.async0[E, Option[A]](_ => {
```

Comment:
Use `iO.flatten(IO.sync( /* IO.now / IO.terminate / IO.fail go here */ ))`

Reply:
Also it should probably expose the full state (`Option[ExitResult[E, A]]`). An additional helper method could check to see if it contains an `A`. Helpers should always be written in terms of the most powerful / generalized version.